Bmx.Idp.Okta/OktaClient.cs

```diff
@@ -46,37 +61,98 @@ public class OktaClient : IIdentityProvider {
 		}
 
 		public async Task<bool> ChallengeMfa( int selectedMfaIndex, string challengeResponse ) {
-			var stateToken = ( (AuthenticateResponseInital)_authenticateResponses[AuthenticationStatus.MFA_REQUIRED] )
-				.StateToken;
-			var mfaFactor =
-				( (AuthenticateResponseInital)_authenticateResponses[AuthenticationStatus.MFA_REQUIRED] ).Embedded
-				.Factors[selectedMfaIndex];
+			var mfaFactor = _oktaMfaFactors[selectedMfaIndex];
 
 			var authResp =
 				await AuthenticateChallengeMfaOkta(
-					new AuthenticateChallengeMfaOptions( mfaFactor.Id, challengeResponse, stateToken ) );
+					new AuthenticateChallengeMfaOptions( mfaFactor.Id, challengeResponse, _oktaStateToken ) );
 
-			_authenticateResponses.Add( AuthenticationStatus.SUCCESS, authResp );
+			_oktaSessionToken = authResp.SessionToken;
 
 			// TODO: Check for auth successes and return state
 			return true;
-			// TODO: Implicitly Call Session here
+		}
+
+		public async Task<string[]> GetAccounts( string accountType ) {
+			// TODO: Use existing session if it exists in ~/.bmx and isn't expired
+			var sessionResp = await CreateSessionOkta( new SessionOptions( _oktaSessionToken ) );
+			AddSession( sessionResp.Id );
+
+			_oktaApps = await GetAccountsOkta( sessionResp.UserId );
+			return _oktaApps.Where( app => app.AppName == accountType ).Select( app => app.Label ).ToArray();
+		}
+
+		public async Task<XmlDocument> GetServiceProviderSaml( int selectedAccountIndex ) {
+			var account = _oktaApps[selectedAccountIndex];
+			var accountPage = await GetAccountOkta( new Uri( account.LinkUrl ) );
+			return ParseAWSSaml( accountPage );
+		}
+
+		private void AddSession( string sessionId ) {
+			CookieContainer.Add( new Cookie( "sid", sessionId, "/", HttpClient.BaseAddress.Host ) );
+		}
+
+		private XmlDocument ParseAWSSaml( string htmlResponse ) {
+			// HTML page is fairly malformed, grab just the <input> with the SAML data for further processing
+			var inputRegexPattern = "<input name=\"SAMLResponse\" type=\"hidden\" value=\".*\"/>";
```

Comment:
A XML or HTML parser is definitely preferred over regex. General good practice reason is that there are less security risks involved. Fun stackoverflow answer: https://stackoverflow.com/a/1732454/2687324.

I think
```suggestion
			var inputRegexPattern = "<input name=\"SAMLResponse\" type=\"hidden\" value=\".*?\"/>";
```
(non-greedy) is what you want here, since you're only looking for one specific HTML tag?

Reply:
Yeah, I had that SO answer in mind when writing this 😛 

Issue was since HTML is not valid XML, .NET Core's built in XML parsing cannot handle it.

Considered using one of the many HTML parsing libs for .NET but decided to use regex since after all there is only one instance of HTML parsing in the whole app (granted that one step is crucial for things to work).

Would you say that makes it a bit more valid to HTML parsing?

Reply:
It seems safe enough here. Could help with keeping the binary size down, since that was a concern noted elsewhere.

Thoughts on using `.*?`?

Reply:
> Thoughts on using .*??

Sounds good to me, implemented.